### PR TITLE
591 : Add support for pagination

### PR DESCRIPTION
**Pull request by rehammuzzamil**
https://github.com/OpenSRP/opensrp-server-web/issues/591

### CODE REVIEW

File: src/main/java/org/opensrp/search/PractitionerSearchBean.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.opensrp.search;
+
+import lombok.Setter;
+import lombok.Getter;
+import lombok.Builder;
+
+@Getter
+@Setter
```
Comment (githengi) [inline src/main/java/org/opensrp/search/PractitionerSearchBean.java:8]: Simplify to `@Data`

File: src/main/java/org/opensrp/service/OrganizationService.java [hunk 2]
```diff
@@ -184,16 +190,17 @@ public void assignLocationAndPlan(String identifier, String jurisdictionId, Stri
...
+		Long planId = planRepository.retrievePrimaryKey(assignedLocationAndPlanSearchBean.getPlanIdentifier());
 
         if (planId == null)
             throw new IllegalArgumentException("Plan not found");
 
-		return organizationRepository.findAssignedLocationsByPlanId(planId);
+        assignedLocationAndPlanSearchBean.setPlanId(planId);
+		return organizationRepository.findAssignedLocationsByPlanId(assignedLocationAndPlanSearchBean);
 
 	}
```
Comment (githengi) [inline src/main/java/org/opensrp/service/OrganizationService.java:205]: To make it easy to understand this method, can we retain the original contract and just add the pagination params. You can package the params back when invoking the repository

File: src/main/java/org/opensrp/service/OrganizationService.java [hunk 3]
```diff
@@ -146,12 +147,17 @@ public void assignLocationAndPlan(String identifier, String jurisdictionId, Stri
...
+		if(assignedLocationAndPlanSearchBean != null && assignedLocationAndPlanSearchBean.getOrganizationIdentifier() != null) {
+			validateIdentifier(assignedLocationAndPlanSearchBean.getOrganizationIdentifier());
+			organization = getOrganization(assignedLocationAndPlanSearchBean.getOrganizationIdentifier());
+		}
 		if (organization == null)
 			throw new IllegalArgumentException("Organization not found");
-		return organizationRepository.findAssignedLocations(organization.getId(),returnFutureAssignments);
+
+		assignedLocationAndPlanSearchBean.setOrganizationId(organization.getId());
+		return organizationRepository.findAssignedLocations(assignedLocationAndPlanSearchBean);
```
Comment (githengi) [inline src/main/java/org/opensrp/service/OrganizationService.java:160]: To make it easy to understand this method, can we retain the original contract and just add the pagination params. You can package the params back when invoking the repository

File: src/main/java/org/opensrp/search/BaseSearchBean.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+	private Integer pageNumber = 0;
+
+	private Integer pageSize = 0;
+
+	private BaseSearchBean.OrderByType orderByType;
+
+	private BaseSearchBean.FieldName orderByFieldName;
+
+	public BaseSearchBean(Integer pageNumber, Integer pageSize, OrderByType orderByType,
```
Comment (githengi) [inline src/main/java/org/opensrp/search/BaseSearchBean.java:28]: This is not needed just add an `@AllArgsConstructor`

**Review by githengi — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensrp/search/PractitionerSearchBean.java (hunk 1 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/search/PractitionerSearchBean.java]: Done

File: src/main/java/org/opensrp/search/AssignedLocationAndPlanSearchBean.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+	private String organizationIdentifier;
+
+	private Long planId;
+
+	private Long organizationId;
+
+	private boolean returnFutureAssignments;
+
+	@Builder
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/search/AssignedLocationAndPlanSearchBean.java]: Noted

File: src/main/java/org/opensrp/repository/postgres/OrganizationRepositoryImpl.java [hunk 6]
```diff
@@ -181,33 +182,42 @@ private void insertOrganizationLocation(Long organizationId, Long jurisdictionId
...
-		Criteria criteria = example.createCriteria().andOrganizationIdEqualTo(organizationId);
-		if (!returnFutureAssignments) {
-			criteria.andFromDateLessThanOrEqualTo(currentDate);
+		Pair<Integer, Integer> pageSizeAndOffset = getPageSizeAndOffsetForAssignedLocAndPlans(assignedLocationAndPlanSearchBean);
+		if(assignedLocationAndPlanSearchBean != null && assignedLocationAndPlanSearchBean.getOrganizationId() != null) {
+			Criteria criteria = example.createCriteria()
+					.andOrganizationIdEqualTo(assignedLocationAndPlanSearchBean.getOrganizationId());
+			if (!assignedLocationAndPlanSearchBean.isReturnFutureAssignments()) {
+				criteria.andFromDateLessThanOrEqualTo(currentDate);
+			}
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/OrganizationRepositoryImpl.java]: I have just added null checks to get `organizationId` and `isReturnFutureAssignments` param values, else the logic remains the same.

File: src/main/java/org/opensrp/repository/postgres/PractitionerRepositoryImpl.java [hunk 7]
```diff
@@ -241,4 +252,22 @@ private Practitioner convert(org.opensrp.domain.postgres.Practitioner pgPractiti
...
+        } else {
+            pageSize = practitionerSearchBean.getPageSize();
+        }
+
+        if (practitionerSearchBean.getPageNumber() != null && practitionerSearchBean.getPageNumber() != 0) {
+            offset = (practitionerSearchBean.getPageNumber() - 1) * pageSize;
+        }
+
+        return Pair.of(pageSize, offset);
+    }
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/PractitionerRepositoryImpl.java:274]: Sure.

File: src/main/java/org/opensrp/repository/postgres/PractitionerRoleRepositoryImpl.java [hunk 8]
```diff
@@ -294,4 +304,21 @@ private boolean isExistingPractitionerRole(Long organizationId, Long practitione
...
+        } else {
+            pageSize = practitionerRoleSearchBean.getPageSize();
+        }
+
+        if (practitionerRoleSearchBean.getPageNumber() != null && practitionerRoleSearchBean.getPageNumber() != 0) {
+            offset = (practitionerRoleSearchBean.getPageNumber() - 1) * pageSize;
+        }
+
+        return Pair.of(pageSize, offset);
+    }
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/PractitionerRoleRepositoryImpl.java:327]: Done

File: src/main/java/org/opensrp/search/PractitionerRoleSearchBean.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.Setter;
+import lombok.Getter;
+import lombok.Builder;
+
+@Getter
+@Setter
+public class PractitionerRoleSearchBean extends BaseSearchBean{
+
+	@Builder
+	public PractitionerRoleSearchBean(Integer pageNumber, Integer pageSize,
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/search/PractitionerRoleSearchBean.java]: Done

File: src/main/java/org/opensrp/search/BaseSearchBean.java (hunk 4 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/search/BaseSearchBean.java]: Done

File: src/main/java/org/opensrp/service/OrganizationService.java (hunk 3 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/service/OrganizationService.java]: Noted.

File: src/main/java/org/opensrp/service/OrganizationService.java (hunk 2 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/service/OrganizationService.java]: Okay, noted.

File: src/main/java/org/opensrp/search/PractitionerSearchBean.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+public class PractitionerSearchBean extends BaseSearchBean{
+
+	@Builder
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/search/PractitionerSearchBean.java]: Noted.

File: src/main/java/org/opensrp/repository/OrganizationRepository.java [hunk 11]
```diff
@@ -67,11 +68,13 @@ void assignLocationAndPlan(Long organizationId, String jurisdictionIdentifier, L
 	 *
 	 * @return list of assigned locations and plans
 	 */
-	List<AssignedLocations> findAssignedLocationsByPlanId(Long planId);
+	List<AssignedLocations> findAssignedLocationsByPlanId(AssignedLocationAndPlanSearchBean assignedLocationAndPlanSearchBean);
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/OrganizationRepository.java:72]: Added @Deprecated annotation. Please verify the logic in `findAssignedLocations` methods @githengi

[the remark this answers is not in the document]
Comment (githengi) [reply src/main/java/org/opensrp/repository/OrganizationRepository.java:72]: Add the method that should be used on a javadoc for this methdo

File: src/main/java/org/opensrp/repository/postgres/OrganizationRepositoryImpl.java [hunk 12]
```diff
@@ -181,33 +183,51 @@ private void insertOrganizationLocation(Long organizationId, Long jurisdictionId
...
+			criteria.andOrganizationIdEqualTo(assignedLocationAndPlanSearchBean.getOrganizationId());
+		}
+		if (!assignedLocationAndPlanSearchBean.isReturnFutureAssignments()) {
 			criteria.andFromDateLessThanOrEqualTo(currentDate);
 		}
-		return organizationLocationMapper.findAssignedlocationsAndPlans(example.getOredCriteria(),
-		    example.getOrderByClause(), currentDate);
+		if (assignedLocationAndPlanSearchBean != null && assignedLocationAndPlanSearchBean.getPlanId() != null) {
+			example.createCriteria().andPlanIdEqualTo(assignedLocationAndPlanSearchBean.getPlanId());
+			example.createCriteria().andFromDateLessThanOrEqualTo(currentDate);
```
Comment (githengi) [inline src/main/java/org/opensrp/repository/postgres/OrganizationRepositoryImpl.java:200]: Remove this

Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/OrganizationRepositoryImpl.java]: Okay

File: src/main/java/org/opensrp/repository/postgres/mapper/custom/xml/CustomOrganizationMapper.xml [hunk 13]
```diff
@@ -141,4 +141,22 @@
...
+                refid="org.opensrp.repository.postgres.mapper.OrganizationMapper.Base_Column_List" />
+        from team.organization o
+        <if test="_parameter != null">
+            <include refid="Example_Where_Clause" />
+        </if>
+        <if test="searchBean.orderByFieldName!=null and searchBean.orderByType!=null">
+            order by o.${searchBean.orderByFieldName} ${searchBean.orderByType}
+        </if>
+        LIMIT #{limit} OFFSET #{offset}
+    </select>
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/mapper/custom/xml/CustomOrganizationMapper.xml]: Done

File: src/main/java/org/opensrp/repository/postgres/mapper/custom/xml/CustomPractitionerRoleMapper.xml [hunk 14]
```diff
@@ -47,4 +47,19 @@
...
+                refid="org.opensrp.repository.postgres.mapper.PractitionerRoleMapper.Base_Column_List" />
+        from team.practitioner_role pr
+        <if test="_parameter != null">
+            <include refid="Example_Where_Clause" />
+        </if>
+        <if test="searchBean.orderByFieldName!=null and searchBean.orderByType!=null">
+            order by ${searchBean.orderByFieldName} ${searchBean.orderByType}
+        </if>
+        LIMIT #{limit} OFFSET #{offset}
+    </select>
```
[the remark this answers is not in the document]
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/mapper/custom/xml/CustomPractitionerRoleMapper.xml]: Done

**Review by githengi — APPROVED** [review]

**Pull request merged**